### PR TITLE
allow for multiple beetmover taskcluster_scope_prefixes

### PR DESCRIPTION
**Pull request by escapewindow**
Currently untested.

### CODE REVIEW

File: beetmoverscript/src/beetmoverscript/task.py
Comment (ahal) [inline beetmoverscript/src/beetmoverscript/task.py:59]: nit: I think in-lining this to the function above is a bit more readable.

**Review by ahal — APPROVED** [review]
Thanks, this looks reasonable to me! I hope this doesn't remove Mihai's review.

**Comment by ahal** [comment]
I'll let you decide whether you want to block on Mihai's review or not. It should still show up in his myqonly notifications.

File: beetmoverscript/src/beetmoverscript/task.py
Comment (ahal) [reply beetmoverscript/src/beetmoverscript/task.py:59]: Another slightly more efficient way of doing this check would be to assert the `len` of scopes is the same before and after the filtering.

File: beetmoverscript/src/beetmoverscript/task.py
Comment (ahal) [inline beetmoverscript/src/beetmoverscript/task.py:64]: nit: Any reason not to use f-strings here? I also tend to like to format lists per-line in error messages, e.g:
```python
scope_str = " - ".join(scopes)
prefix_str = " - ".join(prefixes)
... f"Given scopes:\n  -{scope_str}\n\nAllowed prefixes:\n -{prefix_str}"...
```

though it seems like this probably won't ever happen so, meh. Up to you.

**Comment by escapewindow** [comment]
Re: both review comments: agreed; I mainly copied over [signingscript's code](https://github.com/mozilla-releng/scriptworker-scripts/blob/master/signingscript/src/signingscript/task.py#L105-L126) with minor changes. We might even want to move these helper functions to, say, `scriptworker_client` so we can share them. I'm open to whichever: land as is, address the review comments, move these to `scriptworker_client` and import in both places.

**Comment by MihaiTabara** [comment]
Was chatting with Stefan on adjusting https://github.com/mozilla-mobile/android-components/blob/master/taskcluster/ci/config.yml#L190 and it seemed that we'd only need to land that on `master`, if need be. However, that changes scopes for everything, signing and pushapk included. Tackling the multiple prefixes might remove us from the weeds sooner. Looking at the PR now

File: beetmoverscript/docker.d/worker.yml
Comment (MihaiTabara) [inline beetmoverscript/docker.d/worker.yml:16]: I presume we want `project:mobile:fenix:releng:beetmover` as well here? We'll want Focus APKs published too so ++ on adding Focus too from now.

**Comment by MihaiTabara** [comment]
@escapewindow @ahal for the record, I've noticed we do have a `beetmover-dev-relengworker-firefoxci-mobile-1-1` within our workers so we could also YOLO this in the `dev-beetmoverscript`, provided that ahal's changing his patches in Fenix to point to the `-dex` workes instead of the staging ones.

File: beetmoverscript/docker.d/worker.yml
Comment (MihaiTabara) [reply beetmoverscript/docker.d/worker.yml:16]: I assume we also need to add the Fenix corresponding S3 credentials, something like down below, but that's not blocking this PR - I suppose you want to see whether this PR first is backward compatible? :
```
diff --git a/beetmoverscript/docker.d/worker.yml b/beetmoverscript/docker.d/worker.yml
index f06f089..f4e4905 100644
--- a/beetmoverscript/docker.d/worker.yml
+++ b/beetmoverscript/docker.d/worker.yml
@@ -128,6 +128,27 @@ bucket_config:
             nightly_components: 'maven-nightly-stage-s3-upload-bucket'
         dummy: {}
       'COT_PRODUCT == "mobile" && ENV == "prod"':
+        nightly:
+          credentials:
+            id: { "$eval": "FENIX_NIGHTLY_ID" }
+            key: { "$eval": "FENIX_NIGHTLY_KEY" }
+          buckets:
+            fenix: 'net-mozaws-prod-delivery-archive'
+          url_prefix: 'https://archive.mozilla.org'
+        release:
+          credentials:
+            id: { "$eval": "FENIX_RELEASE_ID" }
+            key: { "$eval": "FENIX_RELEASE_KEY" }
+          buckets:
+            fenix: 'net-mozaws-prod-delivery-archive'
+          url_prefix: 'https://archive.mozilla.org'
+        dep:
+          credentials:
+            id: { "$eval": "FENIX_DEP_ID" }
+            key: { "$eval": "FENIX_DEP_KEY" }
+          buckets:
+            fenix: 'net-mozaws-stage-delivery-archive'
+          url_prefix: 'https://ftp.stage.mozaws.net'
         maven-production:
           credentials:
             id: { "$eval": "MAVEN_ID" }
```

Comment (escapewindow) [reply beetmoverscript/docker.d/worker.yml:16]: > I presume we want `project:mobile:fenix:releng:beetmover` as well here? We'll want Focus APKs published too so ++ on adding Focus too from now.

Oops, good catch :) I mean, I *meant* to proactively add Focus.

**Comment by ahal** [comment]
I thought you were surreptitiously leaving it out to see if I would figure
out to add it later ;)

**Comment by escapewindow** [comment]
Added the fenix prefix, and pushed to dev-beetmoverscript to either give us an error message or create the mobile-1-beetmover-dev pool.

**Comment by MihaiTabara** [comment]
> Added the fenix prefix, and pushed to dev-beetmoverscript to either give us an error message or create the mobile-1-beetmover-dev pool.

Makes sense. Though I think we'd need an actual task scheduled afterwards, given https://github.com/mozilla-releng/k8s-autoscale/blob/master/configs/relengworker-nonprod/config.yml#L28

File: beetmoverscript/docker.d/worker.yml
Comment (escapewindow) [reply beetmoverscript/docker.d/worker.yml:16]: > I assume we also need to add the Fenix corresponding S3 credentials, something like down below, but that's not blocking this PR - I suppose you want to see whether this PR first is backward compatible?

Yeah, I wanted to solve the prefix problem here, and then we can add the bucket creds and such in a followup.

Comment (escapewindow) [reply beetmoverscript/docker.d/worker.yml:16]: Ok, this is on dev-beetmoverscript now, and appears to work. We probably want to add the staging bucket secrets to SOPS and update the worker secrets, and then we can test from staging-fenix. @ahal how much of this do you want to handle? Do you want a chat / writeup / zoom call about this? (@MihaiTabara or I probably need to update SOPS.)

**Comment by ahal** [comment]
I don't mind pushing on the secrets if you think that's useful for me to
know. If so then yes, I'd need some guidance.

 Otherwise I have some work I can do in the meantime.. basically get to the
point where I can see a failure pushing to the bucket due to missing
credentials.

**Comment by MihaiTabara** [comment]
Pushing in a bit my stashed changes for this.

**Comment by ahal** [comment]
Please cc me on the PR/bug (assuming I'm allowed to see it), just so I can
see how it's done.

**Comment by MihaiTabara** [comment]
@escapewindow @ahal I've added a bit more support for this in beetmover. I'm not very happy with the code changes, but they should unblock ahal for now. I have thoughts on how to improve this further but may be good enough for now and tackle in a follow-up more? 

Note: https://gist.github.com/MihaiTabara/594bf50c4000b7909f01db7fa7f1760b is the `task.json` I've used to test these changes.

Note to self:
* fix existing tests
* I still need to add tests for this
* cleanup the schemas, they are a mess right now
* add the corresponding `DEP_ID` and `DEP_KEY` with dummy values for now in SOPS + re-deploy the dev workers

File: beetmoverscript/docker.d/init_worker.sh
Comment (MihaiTabara) [inline beetmoverscript/docker.d/init_worker.sh:69]: Note to self: I need to add these in SOPS  + redeploy dev workers

File: beetmoverscript/docker.d/init_worker.sh
Comment (MihaiTabara) [inline beetmoverscript/docker.d/init_worker.sh:77]: Note to self: I need to add these in SOPS + redeploy dev workers

File: beetmoverscript/docker.d/worker.yml
Comment (MihaiTabara) [reply beetmoverscript/docker.d/worker.yml:16]: Added a follow-up commit with missing changes.

File: beetmoverscript/docker.d/worker.yml
Comment (MihaiTabara) [inline beetmoverscript/docker.d/worker.yml:144]: We're likely getting real credentials only next week but I think for the dev workers, we can totally fake with dummy values in SOPS until then.

File: beetmoverscript/src/beetmoverscript/data/artifactMap_beetmover_task_schema.json
Comment (MihaiTabara) [inline beetmoverscript/src/beetmoverscript/data/artifactMap_beetmover_task_schema.json:57]: This schema was unused since we switched to declarative artifacts so I cleared it and reused it. I don't like at all how schemas look like currently, it's a mess, but this should unblock for now at least.

File: beetmoverscript/src/beetmoverscript/script.py
Comment (MihaiTabara) [inline beetmoverscript/src/beetmoverscript/script.py:113]: @escapewindow curious what your thoughts are here. For Fenix, we'd likely push directly into ~releases, so no `candidates` so it didn't make sense to reuse that logic. On the other hand, the `push-to-candidates` also assumes balrog manifest logic + checksums which we don't really need here for now. So I thought adding a new action scope to reflect that since the `push-to-releases` is an existing different snippet.

File: beetmoverscript/src/beetmoverscript/task.py
Comment (MihaiTabara) [reply beetmoverscript/src/beetmoverscript/task.py:59]: Note to self: @ahal and I spoke about this and we agreed to ignore for now, file an issue and move these all in `scriptworker_client` to benefit in the future.

File: docs/scriptworkers-FAQ.md
Comment (MihaiTabara) [inline docs/scriptworkers-FAQ.md:36]: Adjusting a bit the docs since I found it incomplete - https://scriptworker-scripts.readthedocs.io/en/latest/scriptworkers-FAQ.html

File: beetmoverscript/docker.d/init_worker.sh
Comment (escapewindow) [reply beetmoverscript/docker.d/init_worker.sh:69]: We also need to update the tests to populate fake values here.

File: beetmoverscript/src/beetmoverscript/script.py
Comment (escapewindow) [reply beetmoverscript/src/beetmoverscript/script.py:113]: I'm ok with that. We may want to somehow list which trust domains support direct pushes, and which ones support candidates->releases, so we don't accidentally do the wrong thing in the wrong project.

File: beetmoverscript/docker.d/init_worker.sh
Comment (MihaiTabara) [reply beetmoverscript/docker.d/init_worker.sh:69]: > Note to self: I need to add these in SOPS + redeploy dev workers

I've populated with fake values the `DEP_ID` and `DEP_KEY` values and redeployed (I foced redeploy to skip the tox blocking dep) `dev-mobile-beetmoverworkers` so that they contain the HEAD of this PR. Locally this works against https://gist.github.com/MihaiTabara/594bf50c4000b7909f01db7fa7f1760b, with the caveat of 401 against S3, of course. I have the production credentials from Jeremy, but he's going to follow-up with the staging ones in the next days too. 

I'll address the rest of the leftovers TODOs from this PR when I'm back from PTO, but for now, this should be enough to unblock @ahal's testing to see a beetmover red task. CC @escapewindow

Comment (MihaiTabara) [reply beetmoverscript/docker.d/init_worker.sh:69]: I believe @hneiva has already overwritten the dep keys in SOPS so we should be good on this section.

File: beetmoverscript/docker.d/init_worker.sh
Comment (MihaiTabara) [reply beetmoverscript/docker.d/init_worker.sh:77]: @CC @hneiva I think we'll have to make sure we have these setup properly in SOPS + cloudops-infra. I may have set them to dummy values last time, but if I didn't, we need to make sure there are valid values there.

File: beetmoverscript/docker.d/worker.yml
Comment (MihaiTabara) [reply beetmoverscript/docker.d/worker.yml:16]: I think @hneiva deployed this successfully this week so we should be good here.

Comment (MihaiTabara) [reply beetmoverscript/docker.d/worker.yml:16]: @hneiva It might be worth checking with @sarah-clements as well whether Focus team wants this uploaded to S3 as well in the future? We don't need to block this PR but if we do, it'd be good to double-check the structure so that when that request comes along, it'll be easier to fill in those credentials.

File: beetmoverscript/docker.d/worker.yml
Comment (MihaiTabara) [reply beetmoverscript/docker.d/worker.yml:144]: We did, I think we're good here.

File: beetmoverscript/src/beetmoverscript/script.py
Comment (MihaiTabara) [reply beetmoverscript/src/beetmoverscript/script.py:113]: @escapewindow Resurrecting this thread - so it sounds like you're thinking of some additional check similar to https://github.com/mozilla-releng/scriptworker/blob/master/src/scriptworker/constants.py#L320 somewhere there but adjusted to cot product / actions?  CC @hneiva

Comment (escapewindow) [reply beetmoverscript/src/beetmoverscript/script.py:113]: Yes, and/or baking it into the script config, e.g. iscript's [supported behaviors](https://github.com/mozilla-releng/scriptworker-scripts/blob/master/iscript/config_example.yaml#L15-L20) allowlist.

**Comment by hneiva** [comment]
@escapewindow We should close this PR

**Pull request closed without merging**
